PowerShellTools.HostService/ServiceManagement/Debugging/PowerShellDebuggingService.cs

```diff
@@ -132,6 +144,124 @@ public void SetRunspace(bool overrideExecutionPolicy)
+            if (_callback == null)
+            {
+                _callback = OperationContext.Current.GetCallbackChannel<IDebugEngineCallback>();
+            }
+
+            // Attaching leverages cmdlets introduced in PSv5
+            if (_installedPowerShellVersion < RequiredPowerShellVersionForProcessAttach)
+            {
+                MessageBox.Show(Constants.ProcessAttachVersionError, "Invalid Version", MessageBoxButton.OK, MessageBoxImage.Warning);
+                return;
+            }
+
+            // Enter into to-attach process which will swap out the current runspace
+            _attaching = true;
+            _needToOpen = true;
+            _currentPowerShell = PowerShell.Create();
+            _currentPowerShell.Runspace = _runspace;
+            _currentPowerShell.AddCommand("Enter-PSHostProcess").AddParameter("Id", pid.ToString());
+            _currentPowerShell.Invoke();
+
+            // wait for invoke to finish swapping the runspaces
+            _attachRequestEvent.WaitOne(5000);
+
+            // rehook up the event handling
+            _runspace.Debugger.DebuggerStop += Debugger_DebuggerStop;
+            _runspace.Debugger.BreakpointUpdated += Debugger_BreakpointUpdated;
+            _runspace.StateChanged += _runspace_StateChanged;
+            _runspace.AvailabilityChanged += _runspace_AvailabilityChanged;
+
+            // debug the runspace
+            _currentPowerShell.Runspace = _runspace;
+            _currentPowerShell.Commands.Clear();
+            _currentPowerShell.AddCommand("Debug-Runspace").AddParameter("Id", "1");
+
+            try
+            {
+                _currentPowerShell.Invoke();
+            }
+            catch (RemoteException ex)
+            {
+                // exception is expected if user asks to stop debugging while script is running
+                ServiceCommon.Log("Forced to detach via stop command; " + ex.ToString());
+            }
+        }
+
+        /// <summary>
+        /// Examines a process' modules for powershell.exe which indicates the process should
+        /// be attachable. Will generate an exception if the host process is x86 and the process
+        /// is x64.
+        /// </summary>
+        /// <param name="pid"></param>
+        /// <returns>True if powershell.exe is a module of the process, false otherwise.</returns>
+        public bool IsAttachable(uint pid)
+        {
+            try
+            {
+                var process = Process.GetProcessById((int)pid);
+                ServiceCommon.Log("IsAttachable:" + process.ProcessName + "; id:" + process.Id);
+                foreach (ProcessModule module in process.Modules)
+                {
+                    if (module.ModuleName.Equals("powershell.exe") && _installedPowerShellVersion >= RequiredPowerShellVersionForProcessAttach)
+                    {
+                        return true;
+                    }
+                }
+            }
+            catch (Exception ex)

```

Comment:
I recalled you mentioned that powershell itself is able to debug x64 process from x86 powershell host, if that is true, we should consider to catch the specific exception (for case where x86 host try to attach to x64 process), and return true for that case so that user can still attach to the process(ofcourse in that case we dont have a way to evaluate if it is actually running powershell.exe, but I think it is still way better than existing behavior)?